### ISSUE TITLE
Pulumi could not detect Resource change without code change.

### ISSUE THREAD

**Issue by guitarrapc** [pulumi/pulumi#3664]
## TL;DR

I do expect Pulumi detect resource change even if there are no code change to the resource. However current pulumi won't detect actual resource change if there are no code change to the resource.

## Description

This is not a bug report, rather then a question or feature request.

There are detailed doc about [How Pulumi Works \| Pulumi](https://www.pulumi.com/docs/intro/concepts/how-pulumi-works/).

Reading throught `Language Hosts` and `Deployment Engine` sections, pulumi will trigger when `resource registrations` is detected on `Language Host`.
From my understanding, `resource registrations` is triggered only when there are any code change happens. So it won't check actual provider's resource if no code change happens from Pulumi's last run state. 

Current behavior seems like documented `desired state` is not Actual State expressed via code, but Last Deployed State expressed via code.

> The deployment engine is responsible for computing the set of operations needed to drive the current state of your infrastructure into the desired state expressed by your program.
>
> refer: Deployment Engine section

However there are some cases that Provider or any other will make change to actual resource.
Pulumi won't detect change for this change and Code will differ from actual resource state until code change was made to `changed resource`. 

I feel it's too difficult for users to know there are changed made to resource without Pulumi detect it.
It's very nice if Pulumi detect change even there are no code changes, so that user could detect desired state was broken as soon as possible.

## Expected Behavior

Pulumi detect pulumi manged resource change even if there are no code change made.

## Actual behavior

Pulumi won't detect pulumi manged resource change if there are no code change made.

## Reproduce Steps

* Create Resource via Pulumi, let's say S3 bucket.
* Making a direct change to a resource, let's say S3 Bucket Tag Key from `foo` to `foo-1`.
* Run Pulumi without code change will not detect actual resource change.

define resource.

```csharp
            new Pulumi.Aws.S3.Bucket("hogemoge", new Pulumi.Aws.S3.BucketArgs
            {
                Tags = new Dictionary<string, object>
                {
                    { "foo", "bar" },
                },
            });
```

pulumi up

```
$ pulumi up

     Type                 Name                Plan     Info                                                                 pulumi:pulumi:Stack  aws-sandbox-master           'dotnet build -nologo .' completed successfully
     pulumi:pulumi:Stack  aws-sandbox-master             2 messages                                                         Type                 Name                Plan       Info                                                           +   └─ aws:s3:Bucket     hogemoge            create

Resources:
    + 1 to create

Do you want to perform this update?
> yes
  no
  details

Resources:
    + 1 created
```

resource successfully created.

![image](https://user-images.githubusercontent.com/3856350/70896672-7e71d900-2034-11ea-87a1-d3ab79df7d5f.png)

let's simulate direct change tags `foo` -> `foo-1`.

![image](https://user-images.githubusercontent.com/3856350/70898853-ce529f00-2038-11ea-82a3-0809c0953e44.png)

re-run pulumi without code change will not detect real resource change. (foo -> foo-1)

```
Previewing update (master):
     Type                 Name                Plan     Info                                                                 pulumi:pulumi:Stack  aws-sandbox-master           'dotnet build -nologo .' completed successfully                 
     pulumi:pulumi:Stack  aws-sandbox-master           2 messages                                                           

Resources:
    1 unchanged
```

### Compare same steps with Terraform

terrafom detects actual resource change without code change.

define resource.

```tf
resource "aws_s3_bucket" "foo" {
  bucket = "foo-1234sdfgb"
  tags = {
    foo = "bar"
  }
}
```

terraform apply

```
$ terraform apply

Terraform v0.12.18

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_s3_bucket.foo will be created
  + resource "aws_s3_bucket" "foo" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "foo-1234sdfgb"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags                        = {
          + "foo" = "bar"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }
    }

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

resource successfully created.

![image](https://user-images.githubusercontent.com/3856350/70898806-bbd86580-2038-11ea-9a3a-3ddbf3c811a8.png)

let's simulate direct change tags `foo` -> `foo-1`.

![image](https://user-images.githubusercontent.com/3856350/70898836-c98deb00-2038-11ea-8e47-05b7904fa203.png)

re-run terraform without code change will detect real resource change. (foo -> foo-1)

```
$ terraform apply

Terraform v0.12.18
Configuring remote state backend...
Initializing Terraform configuration...
-----------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_s3_bucket.foo will be updated in-place
  ~ resource "aws_s3_bucket" "foo" {
        acl                         = "private"
        arn                         = "arn:aws:s3:::foo-1234sdfgb"
        bucket                      = "foo-1234sdfgb"
        bucket_domain_name          = "foo-1234sdfgb.s3.amazonaws.com"
        bucket_regional_domain_name = "foo-1234sdfgb.s3.ap-northeast-1.amazonaws.com"
        force_destroy               = false
        hosted_zone_id              = "Z2M4EHUR26P7ZW"
        id                          = "foo-1234sdfgb"
        region                      = "ap-northeast-1"
        request_payer               = "BucketOwner"
      ~ tags                        = {
          + "foo"   = "bar"
          - "foo-1" = "bar" -> null
        }

        versioning {
            enabled    = false
            mfa_delete = false
        }
    }

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

**Comment by lukehoban** [comment]
Thanks for the detailed notes @guitarrapc!

I think this boils down to the fact that Pulumi doesn't do a "refresh" by default.  You can manually refresh with `pulumi refresh` or with `pulumi up --refresh`.  Terraform does this by default, and you have to opt-out.  We have considered changing this default, though there are meaningful downsides to doing so.  If you have a stack that you expect to be frequently updated outside of Pulumi though, you may well want to pass `--refresh` to any `pulumi up` to ensure that you are refreshing from the state in the cloud before attempting to make any changes.

I'm going to close this out as a dupe of  https://github.com/pulumi/pulumi/issues/2247, though if there are any other issues you think we should track here do let me know.

**Comment by guitarrapc** [comment]
Thank you for the detail explanation. It make sense.
